bazel: upgrade to c++23

rockwotj opened the pull request:
🎉

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

rockwotj committed:
build: upgrade to c++23

rockwotj committed:
bazel: bump version

vbotbuildovich commented:
<!--RETRIED FAILURES-->
<!--4e2ee41d-bb23-4c33-b8ae-123d34ff91c3-->
<!--RETRIED FAILURES-->
<!--b602e461-ee2f-496e-898d-fb278b73e1b2-->
<!--RETRIED FAILURES-->
<!--f5d8936d-10ea-4a84-9d7b-335a73b0cf19-->
<!--RETRIED FAILURES-->
<!--5644ebe8-12fe-4fcb-9a73-e74fb7007057-->
<!--RETRIED FAILURES-->
<!--947308c0-785c-4bed-aa99-8144b87da6bd-->
<!--RETRIED FAILURES-->
<!--f1737708-cf9d-4cf0-8d6b-2a02ba1b3051-->
<!--RETRIED FAILURES-->
<!--257eb84b-a166-4619-88b4-5dd152d66144-->
<!--body-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59311#019398d8-97fc-4210-9078-5b8e3484b9b7
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59311#019398dd-57c6-483a-97e8-1ea8de0b7d29
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59311#019398d8-97f9-405e-ae10-b53d134d5b1b
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59345#01939a63-859d-4718-9161-f93e9d6b79dd
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59345#01939a7f-077c-4f3b-8770-bf5a7272fcd6
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59459#0193a67d-b680-4215-a0aa-0158e732b846
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59459#0193a699-d4be-4cdb-9878-caf3d53c72d5
<!--/body-->

vbotbuildovich commented:
<!--NEW FAILURES-->
<!--4749a54a-b8b0-47d0-8ee6-290f3e251f57-->
<!--body-->
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59311#019398dd-57c3-4843-a055-8d4b3e9c4031:
```
"rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.cloud_storage_type=CloudStorageType.S3"
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59311-->
<!--959920f1-215d-4057-bfaa-0491dbaa82e1-->
<!--title-->
## Retry command for [Build#59311](https://buildkite.com/redpanda/redpanda/builds/59311)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":false,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
```
<!--/body-->

rockwotj committed:
cluster: use scoped enum

In C++23 we check for scoped enums in serde, we need to make this enum
a class enum.

rockwotj committed:
cloud_topics: remove unused variable

rockwotj committed:
crypto: use deleter after class is defined

Fixes a compilation error in c++23 where we declare the deleter for a
class that hasn't yet been defined.

rockwotj committed:
serde: correct feature check

When this macro is actually defined, you get an error. Fix the check to
use defined instead

rockwotj commented on CMakeLists.txt:
```diff
@@ -12,7 +12,7 @@ project(redpanda LANGUAGES CXX)
 list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
 
 set(CMAKE_C_STANDARD 11)
-set(CMAKE_CXX_STANDARD 20)
```
I still need to do this in vtools, this is for the public build which isn't maintained.

rockwotj commented:
This is only for Bazel, once this lands I can upgrade vtools build as well

michael-redpanda commented:
This will probably conflict: https://github.com/redpanda-data/redpanda/pull/24473

It pulls a ref that includes your changes

michael-redpanda commented:
PR to use seastar fix in bazel: https://github.com/redpanda-data/redpanda/pull/24473

rockwotj commented:
thanks a bazel test is broken anyway

rockwotj committed:
tracking_allocator: support c++23

There was a new allocate_at_least method added

rockwotj commented:
Fixed an allocator change in cpp23 and rebased to fix conflicts

dotnwat commented on src/v/serde/serde_is_enum.h:
```diff
@@ -17,7 +17,7 @@ using serde_enum_serialized_t = int32_t;
 template<typename T>
 inline constexpr bool serde_is_enum_v =
-#if __has_cpp_attribute(__cpp_lib_is_scoped_enum)
+#if defined(__cpp_lib_is_scoped_enum) && __cpp_lib_is_scoped_enum >= 202011L
   std::is_scoped_enum_v<T>
   && sizeof(std::decay_t<T>) <= sizeof(serde_enum_serialized_t);
 #else
```
why do we need the feature check if this is available in c++23? is it the case that it isn't in libc++?

rockwotj replied:
Because vtools is still c++20

rockwotj replied:
https://github.com/redpanda-data/vtools/pull/3362

dotnwat replied:
makes sense

dotnwat commented on src/v/utils/tracking_allocator.h:
```diff
@@ -121,6 +121,14 @@ class tracking_allocator : public allocator {
         return allocator::allocate(n);
     }
 
+#if __cplusplus >= 202302L
+    [[nodiscard]] std::allocation_result<T*> allocate_at_least(std::size_t n) {
+        auto result = allocator::allocate_at_least(n);
+        _tracker->allocate(result.count * sizeof(T));
+        return result;
+    }
+#endif
```
@bharathv
